src/adaptors/adaptor_common.c

```diff
@@ -300,3 +300,26 @@ char *qd_raw_conn_get_address(pn_raw_connection_t *pn_raw_conn)
         return strndup(buffer, 1024);
     }
 }
+
+int qd_raw_connection_drain_read_write_buffers(pn_raw_connection_t *pn_raw_conn)
+{
+    pn_raw_buffer_t buffs[RAW_BUFFER_BATCH];
+    size_t          n;
+    int             buffers_drained = 0;
+    while ((n = pn_raw_connection_take_written_buffers(pn_raw_conn, buffs, RAW_BUFFER_BATCH))) {
+        for (size_t i = 0; i < n; ++i) {
+            buffers_drained++;
+            qd_adaptor_buffer_t *qd_adaptor_buffer = (qd_adaptor_buffer_t *) buffs[i].context;
+            qd_adaptor_buffer_free(qd_adaptor_buffer);
+        }
+    }
+
+    while ((n = pn_raw_connection_take_read_buffers(pn_raw_conn, buffs, RAW_BUFFER_BATCH))) {
+        for (size_t i = 0; i < n; ++i) {
+            buffers_drained++;
+            qd_adaptor_buffer_t *qd_adaptor_buffer = (qd_adaptor_buffer_t *) buffs[i].context;
```

Comment:
Please add an assert here that fails if any of these read buffers contain valid data. In other words: assert(bufs[i].size == 0)

Zero-cost defensive programming.  Trust no one.  Even me...